### ISSUE TITLE
2.x Write Server Initial Sync

### ISSUE THREAD

**Issue by qiangdavidliu** [Netflix/eureka#477]
When a write server boots up, it should not be marked as UP until after it has synced up with all of it's peers. A corollary of this is that the replication protocol need to support metadata exchanges to indicate to receiver servers that it has received all the information the sender has for the initial sync.

**Comment by tbak** [comment]
# Proposed implementation

Eureka server bootstrapping may happen from different sources, depending on Eureka cluster lifecycle:

* for initial startup, there is no prior data available, so servers should start with empty registries
* for existing cluster, when a new write server is brought up, it should ideally take full registry content from one of its peers
* when full cluster is restarted, or there is other reason why a new server cannot bootstrap itself from a peer, it should load last known registry content from an external source/backup storage. Multiple backup storages may be implemented here.

In the original design, we envisaged that peer bootstrapping would be implemented within replication protocol itself:
* a new node chooses one of its peers, and initiates full registry bootstrap
* the chosen peer sends all its registry content, including local and any other registrations
* one bootstrapping is done, a regular replication mode is enabled

This implementation has a few drawbacks:
* complicates replication protocol, with additional mode of operation
* has complex failover scenario - if bootstrapping from first node fails, we have to restart it from another peer, with which we may have already established replication channel.
* it does not cover bootstrapping from external sources, and how the two would interact

In general, we need an API and some pluggability mechanism where multiple bootstrap sources may be provided. They should be also prioritized, so for example peer bootstrap is always first choice. The proposed API consists of two interfaces:

```java
public class RegistryBootstrapCoordinator {

    @Inject
    public RegistryBootstrapCoordinator(RegistryBootstrapService primaryRegistryBootstrapService,
                                        Set<RegistryBootstrapService> additionalRegistryBootstrapServices,
                                        SourcedEurekaRegistry registry) {
        // Setup bootstrap sources
    }

    @PostConstruct
    public void bootstrap() {
        // Initiate registry upload from the configured bootstrap source
    }
}

public interface RegistryBootstrapService {

     Observable<Void> loadIntoRegistry(SourcedEurekaRegistry<InstanceInfo> registry, Source source);
}
```

Each registry entry provided by bootstrap source is associated with ```Source.BOOTSTRAP```, and is put immediately on eviction queue. If bootstrapping fails, the server is enabled anyway, as eventually the up to date state will be reconstructed from the registering clients. The bootstrap process is guarded by a timeout, which if crossed stops it and enables the normal operation mode of the server.

Peer bootstrapping instead of replication protocol, uses regular interest subscription with full registry fetch, with failover mechanism switching to another available service.

# Server availability status

Within Eureka write cluster we cannot depend on instance registered status, when deciding if a server is up or down. Instead this information has to be provided somehow when connecting to the server. A few options may be envisaged here:
* Embed server availability in a framing protocol itself, discarding all requests until server is up. A remote endpoint would receive a special frame asking it to disconnect.
* Implement application level hello message exchange, and embedded server availability status within it.
* Implement negative acknowledgment replies (NACK), carrying over detailed error status similarly to HTTP protocol. If ```NACK.SERVER_UNAVAILABLE``` reply is received, a connection to the given server should be dropped.

**Comment by NiteshKant** [comment]
What will be the downside of only considering option 3:

> When full cluster is restarted, or there is other reason why a new server cannot bootstrap itself from a peer, it should load last known registry content from an external source/backup storage. Multiple backup storages may be implemented here.

as a "bootstrap" scenario and any other scenario as a usual replication scenario? In other words, what is the usecase of getting the entire registry contents from a source (peer or backup) when other peers are available in the cluster?

AFAIU, the entire registry would be required to assert if it is safe to serve registry contents to clients or not. That assertion will be based on the number of instances in a peer's registry. If this is the only reason for the bootstrap, I propose the following:

- Modify replication protocol to also send; 
 - total instances in the peer registry
 - Number of peers

- Modify replication change notification stream to include a marker when the local registry (owned by that node) is streamed completely.

For a node starting up, it should do the following:

1. Attempt to connect to all peers.
2. Get the number of peers and registry information from all peers.
3. If all other peers are reachable then connect to all of them and wait till all of them have streamed all their local registry content.
4. If any of the peer is unreachable or not ready to serve traffic, connect to the backup store to fetch the last known registry information.

A node would start serving registry information to clients, if it has received the registry contents. We may also choose to have some heuristics around how much of the registry information we should wait for in case nodes are not able to stream all their data before disconnect.

The positive of this approach is that you would not get into the hierarchy of bootstrap sources and would also eliminate the need of full registry fetch from a peer in spite of all peers being available.

>> Instead this information has to be provided somehow when connecting to the server. A few options may be envisaged here

I did not quiet understand the difference between the options here. All the options look like sending a message "Not available for registry fetch" and closing the connection. Since, the protocol is framed, so you would have to send the message as a different frame, isn't?

**Comment by tbak** [comment]
Some key points from our offline discussion:
* bootstrapping from replication channel as a first step is a good thing, but slightly complex to implement. We will not do this now, but possibly some time in the future
* doing full registry fetch from a peer or backup cluster should be fast enough, to not impact server bootstrapping time too much
* we will have single external bootstrap source, so the hierarchy proposed above is not needed

What we will provide now for write server bootstrap is a bootstrapping mechanism from either peer write cluster node or external backup cluster (configuration choice).

**Comment by tbak** [comment]
Implemented by PR #500
